File: client/src/modules/assetmanager/schemas/holding/deal-pipeline.schemas.ts
/**
 * Deal Pipeline Schemas
 *
 * Zod validation schemas for DealPipeline model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/holding_models.py
 * - Schema: /server/apps/assetmanager/schemas/holding_schemas/deal_pipeline_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/holding_subrouters/deal_pipeline_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// Enums
// ==========================================

/**
 * Pipeline priority options - matches backend PipelinePriority enum
 * Backend: class PipelinePriority(str, Enum)
 */
export const PipelinePriorityEnum = z.enum([
  'p1',
  'p2',
  'p3',
  'p4',
  'p5',
]);

/**
 * Pipeline status options - matches backend PipelineStatus enum
 * Backend: class PipelineStatus(str, Enum)
 */
export const PipelineStatusEnum = z.enum([
  'initial_screening',
  'due_diligence',
  'term_sheet',
  'negotiation',
  'closing',
  'closed',
  'passed',
  'rejected',
]);

// ==========================================
// DealPipeline Schema (Full Representation)
// ==========================================

/**
 * DealPipeline schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class DealPipeline(BaseModel)
 */
export const DealPipelineSchema = z.object({
  id: z.number(),
  entity_id: z.number(),
  target_entity_id: z.number().nullable(),
  company_name: z.string().nullable(),
  deal_name: z.string(),
  priority: z.string(),
  status: z.string(),
  round_type: z.string(),
  sector: z.string(),

  // Financial Details
  target_raise: z.number().nullable(),
  pre_money_valuation: z.number().nullable(),
  post_money_valuation: z.number().nullable(),
  expected_ownership: z.number().nullable(),
  investment_amount: z.number().nullable(),
  is_lead_investor: z.boolean(),
  other_investors: z.string().nullable(),

  // Dates
  first_contact_date: z.string().nullable(),
  last_interaction_date: z.string().nullable(),
  next_meeting_date: z.string().nullable(),
  expected_close_date: z.string().nullable(),

  // Notes & Analysis
  investment_thesis: z.string().nullable(),
  key_risks: z.string().nullable(),
  due_diligence_notes: z.string().nullable(),
  next_steps: z.string().nullable(),
  rejection_reason: z.string().nullable(),
  notes: z.string().nullable(),

  // Assignment
  assigned_to_id: z.number().nullable(),

  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(), // ISO datetime string from backend
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new deal pipeline entry (POST)
 * Excludes: id, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class DealPipelineCreate(BaseModel)
 */
export const CreateDealPipelineSchema = z.object({
  entity_id: z.number(),
  deal_name: z.string(),
  priority: PipelinePriorityEnum,
  round_type: z.string(),
  sector: z.string(),
  status: PipelineStatusEnum.default('initial_screening'),
  is_lead_investor: z.boolean().optional(),
  target_entity_id: z.number().nullable().optional(),
  company_name: z.string().nullable().optional(),

  // Financial Details
  target_raise: z.number().nullable().optional(),
  pre_money_valuation: z.number().nullable().optional(),
  post_money_valuation: z.number().nullable().optional(),
  expected_ownership: z.number().nullable().optional(),
  investment_amount: z.number().nullable().optional(),
  other_investors: z.string().nullable().optional(),

  // Dates
  first_contact_date: z.string().nullable().optional(),
  last_interaction_date: z.string().nullable().optional(),
  next_meeting_date: z.string().nullable().optional(),
  expected_close_date: z.string().nullable().optional(),

  // Notes & Analysis
  investment_thesis: z.string().nullable().optional(),
  key_risks: z.string().nullable().optional(),
  due_diligence_notes: z.string().nullable().optional(),
  next_steps: z.string().nullable().optional(),
  rejection_reason: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),

  // Assignment
  assigned_to_id: z.number().nullable().optional(),
});

/**
 * Schema for updating a deal pipeline entry (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class DealPipelineUpdate(BaseModel)
 */
export const UpdateDealPipelineSchema = z.object({
  entity_id: z.number().nullable().optional(),
  target_entity_id: z.number().nullable().optional(),
  company_name: z.string().nullable().optional(),
  deal_name: z.string().nullable().optional(),
  priority: PipelinePriorityEnum.nullable().optional(),
  status: PipelineStatusEnum.nullable().optional(),
  round_type: z.string().nullable().optional(),
  sector: z.string().nullable().optional(),

  // Financial Details
  target_raise: z.number().nullable().optional(),
  pre_money_valuation: z.number().nullable().optional(),
  post_money_valuation: z.number().nullable().optional(),
  expected_ownership: z.number().nullable().optional(),
  investment_amount: z.number().nullable().optional(),
  is_lead_investor: z.boolean().nullable().optional(),
  other_investors: z.string().nullable().optional(),

  // Dates
  first_contact_date: z.string().nullable().optional(),
  last_interaction_date: z.string().nullable().optional(),
  next_meeting_date: z.string().nullable().optional(),
  expected_close_date: z.string().nullable().optional(),

  // Notes & Analysis
  investment_thesis: z.string().nullable().optional(),
  key_risks: z.string().nullable().optional(),
  due_diligence_notes: z.string().nullable().optional(),
  next_steps: z.string().nullable().optional(),
  rejection_reason: z.string().nullable().optional(),
  notes: z.string().nullable().optional(),

  // Assignment
  assigned_to_id: z.number().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type PipelinePriority = z.infer<typeof PipelinePriorityEnum>;
export type PipelineStatus = z.infer<typeof PipelineStatusEnum>;
export type DealPipeline = z.infer<typeof DealPipelineSchema>;
export type CreateDealPipeline = z.infer<typeof CreateDealPipelineSchema>;
export type UpdateDealPipeline = z.infer<typeof UpdateDealPipelineSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single deal pipeline entry
 * Backend equivalent: class DealPipelineResponse(BaseModel)
 */
export type DealPipelineResponse = {
  success: boolean;
  data?: DealPipeline;
  error?: string;
};

/**
 * Response containing multiple deal pipeline entries
 * Backend equivalent: class DealPipelinesResponse(BaseModel)
 */
export type DealPipelinesResponse = {
  success: boolean;
  data?: DealPipeline[];
  error?: string;
};

File: server/apps/assetmanager/schemas/holding_schemas/deal_pipeline_schemas.py
"""
DealPipeline Schemas

Pydantic schemas for the DealPipeline model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from datetime import date as date_type
from enum import Enum

# ==========================================
# Enum Types
# ==========================================

class PipelinePriority(str, Enum):
    """Pipeline priority options"""
    P1 = "p1"
    P2 = "p2"
    P3 = "p3"
    P4 = "p4"
    P5 = "p5"

class PipelineStatus(str, Enum):
    """Pipeline status options"""
    INITIAL_SCREENING = "initial_screening"
    DUE_DILIGENCE = "due_diligence"
    TERM_SHEET = "term_sheet"
    NEGOTIATION = "negotiation"
    CLOSING = "closing"
    CLOSED = "closed"
    PASSED = "passed"
    REJECTED = "rejected"

# ==========================================
# DealPipeline Schema (Full Representation)
# ==========================================

class DealPipeline(BaseModel):
    """DealPipeline schema - full representation"""
    id: int
    entity_id: int = Field(description="Associated entity ID")
    target_entity_id: int | None = Field(None, description="Target entity ID")
    company_name: str | None = Field(None, description="Company name")
    deal_name: str = Field(description="Deal name")
    priority: str = Field(description="Priority level (p1-p5)")
    status: str = Field(description="Pipeline status")
    round_type: str = Field(description="Round type")
    sector: str = Field(description="Sector")

    # Financial Details
    target_raise: float | None = Field(None, description="Target raise amount")
    pre_money_valuation: float | None = Field(None, description="Pre-money valuation")
    post_money_valuation: float | None = Field(None, description="Post-money valuation")
    expected_ownership: float | None = Field(None, description="Expected ownership percentage")
    investment_amount: float | None = Field(None, description="Investment amount")
    is_lead_investor: bool = Field(description="Whether lead investor")
    other_investors: str | None = Field(None, description="Other investors")

    # Dates
    first_contact_date: date_type | None = Field(None, description="First contact date")
    last_interaction_date: date_type | None = Field(None, description="Last interaction date")
    next_meeting_date: date_type | None = Field(None, description="Next meeting date")
    expected_close_date: date_type | None = Field(None, description="Expected close date")

    # Notes & Analysis
    investment_thesis: str | None = Field(None, description="Investment thesis")
    key_risks: str | None = Field(None, description="Key risks")
    due_diligence_notes: str | None = Field(None, description="Due diligence notes")
    next_steps: str | None = Field(None, description="Next steps")
    rejection_reason: str | None = Field(None, description="Rejection reason")
    notes: str | None = Field(None, description="Additional notes")

    # Assignment
    assigned_to_id: int | None = Field(None, description="Assigned user ID")

    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class DealPipelineCreate(BaseModel):
    """Schema for creating a new deal pipeline entry"""
    # Required fields
    entity_id: int = Field(description="Associated entity ID")
    deal_name: str = Field(min_length=1, max_length=255, description="Deal name")
    priority: PipelinePriority = Field(description="Priority level (p1-p5)")
    round_type: str = Field(max_length=50, description="Round type")
    sector: str = Field(max_length=50, description="Sector")

    # Fields with defaults
    status: PipelineStatus = Field(default=PipelineStatus.INITIAL_SCREENING, description="Pipeline status")
    is_lead_investor: bool = Field(default=False, description="Whether lead investor")

    # Optional fields
    target_entity_id: int | None = Field(None, description="Target entity ID")
    company_name: str | None = Field(None, max_length=255, description="Company name")

    # Financial Details
    target_raise: float | None = Field(None, description="Target raise amount")
    pre_money_valuation: float | None = Field(None, description="Pre-money valuation")
    post_money_valuation: float | None = Field(None, description="Post-money valuation")
    expected_ownership: float | None = Field(None, description="Expected ownership percentage")
    investment_amount: float | None = Field(None, description="Investment amount")
    other_investors: str | None = Field(None, description="Other investors")

    # Dates
    first_contact_date: date_type | None = Field(None, description="First contact date")
    last_interaction_date: date_type | None = Field(None, description="Last interaction date")
    next_meeting_date: date_type | None = Field(None, description="Next meeting date")
    expected_close_date: date_type | None = Field(None, description="Expected close date")

    # Notes & Analysis
    investment_thesis: str | None = Field(None, description="Investment thesis")
    key_risks: str | None = Field(None, description="Key risks")
    due_diligence_notes: str | None = Field(None, description="Due diligence notes")
    next_steps: str | None = Field(None, description="Next steps")
    rejection_reason: str | None = Field(None, description="Rejection reason")
    notes: str | None = Field(None, description="Additional notes")

    # Assignment
    assigned_to_id: int | None = Field(None, description="Assigned user ID")

class DealPipelineUpdate(BaseModel):
    """Schema for updating a deal pipeline entry"""
    entity_id: int | None = None
    target_entity_id: int | None = None
    company_name: str | None = Field(None, max_length=255)
    deal_name: str | None = Field(None, min_length=1, max_length=255)
    priority: PipelinePriority | None = None
    status: PipelineStatus | None = None
    round_type: str | None = Field(None, max_length=50)
    sector: str | None = Field(None, max_length=50)

    # Financial Details
    target_raise: float | None = None
    pre_money_valuation: float | None = None
    post_money_valuation: float | None = None
    expected_ownership: float | None = None
    investment_amount: float | None = None
    is_lead_investor: bool | None = None
    other_investors: str | None = None

    # Dates
    first_contact_date: date_type | None = None
    last_interaction_date: date_type | None = None
    next_meeting_date: date_type | None = None
    expected_close_date: date_type | None = None

    # Notes & Analysis
    investment_thesis: str | None = None
    key_risks: str | None = None
    due_diligence_notes: str | None = None
    next_steps: str | None = None
    rejection_reason: str | None = None
    notes: str | None = None

    # Assignment
    assigned_to_id: int | None = None

# ==========================================
# Response Types
# ==========================================

class DealPipelineResponse(BaseModel):
    """Response containing a single deal pipeline entry"""
    success: bool
    data: DealPipeline | None = None
    error: str | None = None

class DealPipelinesResponse(BaseModel):
    """Response containing multiple deal pipeline entries"""
    success: bool
    data: list[DealPipeline] | None = None
    error: str | None = None
